### ISSUE TITLE
Extrinsic with `utility.batch` is decoded incorrectly

### ISSUE THREAD

**Issue by ChinmayPatel** [centrifuge/go-substrate-rpc-client#120]
This issue was identified while solving the issue at https://github.com/centrifuge/go-substrate-rpc-client/issues/118 . 

# Bug: 
User is not able to correctly decode the `utility.batch` extrinsic.

# Steps to Reproduce

For simplicity, I've created a Westend transaction @ https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwestend-rpc.polkadot.io#/explorer/query/0xf423599fa7cabb680b19733b5ea0e9676ead737eb38c2bd5ceffcb50a0a950d4 

This transaction has a `utility.batch` extrinsic. The batch contains:
1. a `system.remark` call. 
2. a `balance.transfer` call 

## Correctly decoded using javascript

```
const wsProvider = new WsProvider('wss://westend-rpc.polkadot.io');
const api = await ApiPromise.create({ provider: wsProvider });

const block_hash = '0xf423599fa7cabb680b19733b5ea0e9676ead737eb38c2bd5ceffcb50a0a950d4';

const block = await api.rpc.chain.getBlock(block_hash);

block.block.extrinsics.forEach(ext => {
  console.log("Call Index: ", ext.callIndex[0], ' ', ext.callIndex[1])
});
```
Outputs: 
```
Call Index: 2 0
Call Index: 16 0
```

## Incorrectly decoded using Go

```
api, err := gsrpc.NewSubstrateAPI("wss://westend-rpc.polkadot.io")
if err != nil {
  panic(err)
}

var blockHashString = "0xf423599fa7cabb680b19733b5ea0e9676ead737eb38c2bd5ceffcb50a0a950d4"
blockHash, err := types.NewHashFromHexString(blockHashString)
if(err != nil) {
  panic(err)
}

// Get the block
block, err := api.RPC.Chain.GetBlock(blockHash)
if(err != nil){
  panic(err)
}

meta, err := api.RPC.State.GetMetadata(blockHash)
if err != nil {
  panic(err)
}

for i, ext := range block.Block.Extrinsics {
  fmt.Println("EXT's CALL INDEX ", ext.Method.CallIndex)
}
```
Outputs: 
```
EXT's CALL INDEX {2 0}
EXT's CALL INDEX {61 229}  (Not printed note: this is where the problem is}
``` 

As you can see the second extrinsic's call index is `{61 229}` where it should be `{16 0}`.

# Ask
Both call indexes should be the same. Can you tell us why is this the case?

**Comment by ChinmayPatel** [comment]
Hi @vedhavyas, would you be able to point me in the right direction? If I should talk to someone else, please let me know.

**Comment by vedhavyas** [comment]
Hey @ChinmayPatel
Will investigate and get back

**Comment by ChinmayPatel** [comment]
Thanks @vedhavyas . Looking forward to your response.

**Comment by ChinmayPatel** [comment]
FYI, using `wss://westend-rpc.polkadot.io` for connecting with Polkadot.

**Comment by vedhavyas** [comment]
@ChinmayPatel Seems like this is happening due to the Address decoding on the Extrinsic. The way Go decodes an Address depends on how the address is encoded. If indices are used as a lookup on the chain, then we decode to Account Index else decode it to 32 bytes AccountID. 

The library by default assumes Indices are used. There is a handier function `types.SetSerDeOptions(types.SerDeOptionsFromMetadata(metadata))` to check if the indices pallet is included in the runtime. This works with most of the cases since an indices pallet is used for lookups and if there is no indices pallet, we fallback to decoding 32 bytes of accountID. 

In the `Westend` case, while indices pallet is enabled, it is not used as lookup as shown here in their runtime for reasons unknown.
https://github.com/paritytech/polkadot/blob/master/runtime/westend/src/lib.rs#L130
Hence, this is failing silently.

In centrifuge case, Indices pallet is included and we use that for account lookup as well as shown here
https://github.com/centrifuge/centrifuge-chain/blob/master/runtime/src/lib.rs#L168
Since this is the default, there nothing to be set.

**How does this work in Javascript?**
This works because a few chain type overrides are already set in the lib - 
Westend: https://github.com/polkadot-js/api/blob/master/packages/types-known/src/spec/westend.ts#L11
Centrifuge: https://github.com/polkadot-js/api/blob/master/packages/types-known/src/spec/centrifuge-chain.ts#L11
Hence, this works out of the box.

TLDR: to decode the extrinsic properly for Westend, set no indices before fetching the block as follows
```
types.SetSerDeOptions(types.SerDeOptions{NoPalletIndices: true})
blk, err := api.RPC.Chain.GetBlock(bh)
```

This should give the same result as Javascript
```
{2 0}
{16 0} 
```

**Comment by vedhavyas** [comment]
Closing this issue as the library is working as intended. Please open again if this is not solved

**Comment by ChinmayPatel** [comment]
Going to give this a try today. Thanks for the prompt response. 👍

**Comment by ChinmayPatel** [comment]
That works! Thank you!
